packages/block-library/src/heading/edit.js

```diff
@@ -37,6 +41,33 @@ function HeadingEdit( {
 		style,
 	} );
 
+	// Initially set anchor for headings that have content but no anchor set.
+	// This is used when transforming a block to heading, or for legacy anchors.
+	useEffect( () => {
+		if ( ! anchor && content ) {
+			setAttributes( {
+				anchor: generateAnchor( clientId, content, allHeadingAnchors ),
+			} );
+		}
+		allHeadingAnchors[ clientId ] = anchor;
```

Comment:
```suggestion
		setAnchor( clientId, anchor );
```

If you decide to move the cache variable will need a new `setAnchor` method.